internal/cmd/analyze/analyze.go

```diff
@@ -0,0 +1,622 @@
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package analyze
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/open-policy-agent/opa/ast"
+)
+
+// analyzeResult holds the result of rule analysis that groups the variables
+// in a rule into one-value and many-value buckets.
+// one-value: Variable can have only one value
+// many-value: Variable can be assigned multiple values
+type analyzeResult struct {
+	Raw          *ast.Rule
+	Rule         string
+	OneValue     map[string]ast.Value
+	ManyValue    map[string]string
+	RevManyValue map[string]string
+	Error        error
+	Missing      []string
+	Iterates     bool
+}
+
+// analyzeExpression holds the result of expression analysis
+type analyzeExpression struct {
+	Raw            *ast.Expr
+	Count          string
+	Time           string
+	TimeComplexity string
+}
+
+// RuntimeComplexityResult holds the result for runtime complexity of rules
+type RuntimeComplexityResult struct {
+	Result  map[string][]string
+	Missing map[string][]string
+}
+
+var (
+	varRegexp        = regexp.MustCompile("^[[:alpha:]_][[:alpha:][:digit:]_]*$")
+	virtualDocRegexp = regexp.MustCompile(`{[^}]*}`)
+	underscoreCount  = 0
+
+	varVisitorParams = ast.VarVisitorParams{
+		SkipRefCallHead: true,
+		SkipRefHead:     true,
+	}
+)
+
+// CalculateRuntimeComplexity returns the runtime complexity for the rules
+// in the given module
+func CalculateRuntimeComplexity(m *ast.Module) *RuntimeComplexityResult {
+	results := []*analyzeResult{}
+	finalResult := RuntimeComplexityResult{}
+	finalResult.Result = make(map[string][]string)
+	finalResult.Missing = make(map[string][]string)
+
+	// Walk over the rules in the module and record the variables in them.
+	// Categorize those variables into 2 groups:
+	// 1. Variables which have only one value. eg. x := 1
+	// 2. Variables that can be assigned multiple values. eg. p[x]
+	ast.WalkRules(m, func(r *ast.Rule) bool {
+		result := analyzeResult{}
+		result.Raw = r
+		result.Rule = r.Head.Name.String()
+		result.OneValue = make(map[string]ast.Value)
+		result.ManyValue = make(map[string]string)
+		result.RevManyValue = make(map[string]string)
+		seen := make(map[string]bool)
+		for _, expr := range r.Body {
+			analyzeExpr(expr, seen, &result)
+		}
+		results = append(results, &result)
+
+		return false
+	})
+
+	resultsPre := runtimeComplexityRule(m, results, finalResult.Missing)
+	finalResult.Result = postProcessComplexityResults(resultsPre)
+	return &finalResult
+}
+
+// postProcessComplexityResults walks through the complexity results and
+// replaces results with rule names with their actual complexity results
+//
+// Example:
+// Rule: allow  Complexity: O(foo)
+// Rule: foo    Complexity: O(input.bar)
+//
+// becomes
+//
+// Rule: allow  Complexity: O(input.bar)
+// Rule: foo    Complexity: O(input.bar)
+func postProcessComplexityResults(preRes map[string][]string) map[string][]string {
```

Comment:
NOTE: I haven't reviewed the post processing in-depth yet. I'd like to come back to this after addressing the other comments.